designs/EmbeddedReadmesTechSpec.md

```diff
@@ -0,0 +1,102 @@
+* [ManifestMetadata](https://github.com/NuGet/NuGet.Client/blob/64ed0cb4054226f6060752757d29c50287b312b3/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/ManifestMetadata.cs#L18)
+* [PackageBuilder](https://github.com/NuGet/NuGet.Client/blob/64ed0cb4054226f6060752757d29c50287b312b3/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs#L25)
+
+In order to transmit the `Readme` value from the source file (*nuspec* or *csproj*) to the final output *nuspec* file, the appropriate assignment statements are needed at the relevant locations.
+
+After this, a `ManifestMetadata` object is used to create the output *nuspec* file that is included in the nupkg. For this step, we need to edit the [ToXElement](https://github.com/NuGet/NuGet.Client/blob/64ed0cb4054226f6060752757d29c50287b312b3/src/NuGet.Core/NuGet.Packaging/PackageCreation/Xml/PackageMetadataXmlExtensions.cs#L32) method in PackageMetadataXmlExtensions.cs to include this line:
+
+```
+AddElementIfNotNull(elem, ns, "readme", metadata.Readme);
+```
+
+**If no Readme property is present in the source file, but a readme.md file exists in the base directory**
+
+Once the properties and files have been processed, if the PackageBuilder object's Readme property is null, we need to look through the base directory for a "readme.md" file. If such a file exists, we must include it in the package and and populate the Readme property with "readme.md".
+
+```
+public void CheckForReadme(string basePath)
+{
+    string readmeName = "readme.md";
+    List<PhysicalPackageFile> searchFiles = ResolveSearchPattern(basePath, readmeName, string.Empty, false).ToList();
+
+    if (searchFiles.Any())
+    {
+        Readme = readmeName;
+
+        if (!Files.Contains(searchFiles[0]))
+        {
+            Files.Add(searchFiles[0]);
+        }
+    }
+}
+```
+
+This CheckForReadme method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) is called [here](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs#L725) and [here](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs#L726), based on whether the package is being built using a *nuspec* or *csproj* file.
+
+**Validation and Errors**
+
+To validate the Readme property and the corresponding file when `pack` is called, the ValidateReadmeFile method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) checks that:
+* The file extension matches a Markdown file
+* The file specified in the Readme property exists in the package's files
+* The file can be opened
```

Comment:
Can you clarify what opening a file means in this case?

Reply:
This was meant to cover any other exceptions thrown while attempting to open the embedded readme file, but I've now removed it from the spec in order to be more clear about the validation goals.